src/server/container_utils.h

```diff
@@ -0,0 +1,286 @@
+  // Container stores all values as integral values (int).
+  INTEGRAL,
+  // Container might store strings and integers.
+  MIXED
+};
+
+// StorageType is the respective C++ type of the container internals
+// to accommodate any of its values in the most optimal way.
+// STRING = std::string
+// INTEGRAL = int64_t
+// MIXED = std::variant<std::string, int64_t>
+template <StorageKind K>
+using StorageType =
+    std::conditional_t<K == StorageKind::MIXED, std::variant<std::string, int64_t>,
+                       std::conditional_t<K == StorageKind::STRING, std::string, int64_t>>;
+
+// IsContainer returns true if the iterator points to a container type.
+inline bool IsContainer(const PrimeValue& pv) {
+  unsigned type = pv.ObjType();
+  return (type == OBJ_LIST || type == OBJ_SET || type == OBJ_ZSET);
+}
+
+inline StorageKind GetStorageKind(const PrimeValue& pv) {
+  unsigned type = pv.ObjType();
+  unsigned encoding = pv.Encoding();
+  switch (type) {
+    case OBJ_SET:
+      if (encoding == kEncodingIntSet)
+        return StorageKind::INTEGRAL;
+      else
+        return StorageKind::STRING;
+    case OBJ_ZSET:
+      if (pv.AsRObj()->encoding == OBJ_ENCODING_SKIPLIST)
+        return StorageKind::STRING;
+      else
+        return StorageKind::MIXED;
+    default:
+      return StorageKind::MIXED;
+  };
+}
+
+// Create empty quicklistEntry
+inline quicklistEntry QLEntry() {
+  quicklistEntry res{.quicklist = NULL,
+                     .node = NULL,
+                     .zi = NULL,
+                     .value = NULL,
+                     .longval = 0,
+                     .sz = 0,
+                     .offset = 0};
+  return res;
+}
+
+// Iterate over all values and call func(val). Iteration stops as soon
+// as func return false. Returns true if it successfully processed all elements
+// without stopping.
+template <typename F>
+bool IterateList(const PrimeValue& pv, F&& func, long start = 0, long end = -1) {
+  quicklist* ql = static_cast<quicklist*>(pv.RObjPtr());
+  long llen = quicklistCount(ql);
+  if (end < 0 || end >= llen)
+    end = llen - 1;
+
+  quicklistIter* qiter = quicklistGetIteratorAtIdx(ql, AL_START_HEAD, start);
+  quicklistEntry entry = QLEntry();
+  long lrange = end - start + 1;
+
+  bool success = true;
+  while (success && quicklistNext(qiter, &entry) && lrange-- > 0) {
+    if (entry.value) {
+      success = func(std::string(reinterpret_cast<char*>(entry.value), entry.sz));
+    } else {
+      if constexpr (std::is_invocable_v<F, long long>) {
+        success = func(entry.longval);
+      } else {
+        success = func(absl::StrCat(entry.longval));
+      }
+    }
+  }
+  quicklistReleaseIterator(qiter);
+  return success;
+}
+
+// Iterate over all values and call func(val). If func returns a bool and
+// returns false, iteration stops. func might also return void.
+template <typename F, typename RetV = std::invoke_result_t<F, std::string&&>>
+RetV IterateSet(const PrimeValue& pv, F&& func) {
```

Comment:
This would mean we promote everything to a string. This would also mean dropping support for Integral container types. However given that integer sets are limited to small sizes this would probably not make a significant impact altogether.

Reply:
ah, I missed this.

Reply:
though you use generics but then there is a run-time dispatch inside.